Covariates

edarin opened the pull request:
First version in `output\covariates\final\ua_covariates_oblast.csv`. 
The individual files are in `output\covariates\interim`.

edarin committed:
prep siren covariate

edarin committed:
update acled

edarin committed:
adapt deepstate

edarin committed:
download warfire

edarin committed:
prep warfires

edarin committed:
convert py into r for cov processing

edarin committed:
test pwtt for ukraine

edarin committed:
Merge remote-tracking branch 'upstream/dev' into covariates

edarin committed:
rename pyidp into py_helpers. remove functions not needed in this repo

edarin committed:
transfer py setup to py_helpers

edarin committed:
run pwtt for the entire period and extract monthly value

edarin committed:
Merge branch '58-standardise-model-inputs' into covariates

edarin committed:
Merge remote-tracking branch 'upstream/dev' into covariates

edarin committed:
conform siren to master_index

edarin committed:
process pwtt

edarin committed:
prep war fires

edarin committed:
process acled and standardise outcome

edarin committed:
add imputation = 0 when NA

edarin committed:
remove imputation

edarin committed:
create occupied cov from deepstate

edarin committed:
add country to helper

edarin committed:
aggregate by t occupied

edarin committed:
circumvent one raion in oblast

edarin committed:
rename scripts according to priority

edarin committed:
combine all standardised cov in one

edarin committed:
Merge branch '73-move-location-filtering' into covariates

doug-leasure commented on src/covariates/collection/pwtt.py:
```diff
@@ -0,0 +1,110 @@
+
+
+# data input
+boundaries_oblast_path = in_dir / 'COD-AB' / 'ukr_admbnda_sspe_20230201_SHP'/'ukr_admbnda_adm1_sspe_20230201.shp'
```
Should refer to:
```
data_dir = os.path.join(os.getenv('repo_dir'), 'data')
boundaries_oblast_path = os.path.join(data_dir, 'cod-ab', ...)
```

doug-leasure commented on src/covariates/collection/sirens.R:
```diff
@@ -0,0 +1,122 @@
+  unnest_longer(col = collection_date)
+
+# aggregate sirens per oblast
+# Create a dictionary to map oblast names to i_name names
```
It might be more resilient to map these to `i_key` instead. I think `i_names` is currently using Meta oblast names, which include some Russian spellings of Ukrainian oblast names. We may want to switch to the UN names. This name matching can also be sensitive to special characters.

edarin replied:
but that's the point here. we have names from sirens that are not matching with geo_name so i create the mapping dictionnary?

doug-leasure replied:
If we change the i_names that we are using, then the map will break. I think we will probably change the i_name, but we will not change the i_key.

doug-leasure replied:
We can leave this as it is for now, but may need to revisit later if we modify the i_names.  It's up to you, @edarin

edarin replied:
Oh I understand. you mean having the oblast_map linking `oblast` to `i_key` rather than `oblast` to `i_name`?
I think both are similarly fragile/robust as currently i_name is defined as geo_name and i_key as geo_key. In our settings they are stable because of Meta stability. in other settings, i think we want the i_key <-> i_name stable too, probably from cod-ab. I dont think I should spend time here to replace the i_name to the i_key.

doug-leasure commented on src/covariates/collection/pwtt.py:
```diff
@@ -0,0 +1,110 @@
+from py_helpers.PWTT.code import pwtt
+
+# parameters
+ggdrive = Path('H:/My Drive/')
```
Is it possible to point this to our network drive instead of a personal Google Drive, so everyone can run the code?

If a mounted Google Drive is required by Ollie's code, then lets add a google_dir environment variable, so this works on everyone's machines.

edarin replied:
You need a mounted google drive that is associated with your google earth engine account. google_dir is a nice idea, thanks!

doug-leasure replied:
I moved the Google Drive references to the .env in commit a623e33

doug-leasure commented on data/COD-PS/population_baseline.csv:
This folder should be lower case `./cod-ps/` to match the model pipeline (e.g. `10_create_master_index.R` and `1_mcmc.R`). The lower case also matches the standard already set by the existing folder `./data/cod-ab` which the dashboard uses.

doug-leasure commented on src/1_pop_data/10_create_master_index.R:
A lot of these changes were already committed to other branches and merged with dev. Not sure why it is necessary to recommit them here? 

It's not a problem so long as we aren't undoing any changes made elsewhere.

doug-leasure commented on src/covariates/collection/acled.py:
```diff
@@ -4,55 +4,144 @@
 import os
```
I cannot load the `net_friction` module

edarin replied:
see commit: https://github.com/OxfordDemSci/UkraineNowPop/commit/1b3f333040273fc4a7fdba109ca01a3f386f9e7f

doug-leasure replied:
Nice idea to add code that clones this repo. 

Do we have a requirements.txt for net_friction? It does not work for me because I do not have the dependencies installed in my venv.

doug-leasure replied:
I adapted this solution (176fddb) following the installation instructions from GISRede's repo to use `pip install` which also installs dependencies.

edarin committed:
adapt covariates scripts to the repo_dir

doug-leasure committed:
lower case cod-ps directory

doug-leasure commented on src/covariates/collection/deepstate.py:
```diff
@@ -169,9 +215,66 @@ def process_item(args):
 
 # XXX This is slow, beware
 # Don't worry about weird text output below ("Processing ...")
 processed = dispatch(files)
```
FYI - A couple of `files` returned errors here, but it didn't crash the process.  Probably okay, but just wanted to note. 

```
>>>     processed = dispatch(files)
Processing all 1185 items...

(Processing 663(Processing 550
ERROR 1: sqlite3_exec(CREATE TABLE "occupied_2022-11-09" ( "fid" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "geom" MULTIPOLYGON, "property" TEXT, "date" TEXT)) failed: table "occupied_2022-11-09" already exists
ERROR 1: gdal_sqlite_rtree_bl_from_feature_table() failed with table "rtree_occupied_2022-11-09_geom" already exists

(Processing 814(Processing 791
FeatureError: Could not add feature to layer at index 0: sqlite3_exec(CREATE TABLE "occupied_2022-11-09" ( "fid" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "geom" MULTIPOLYGON, "property" TEXT, "date" TEXT)) failed: table "occupied_2022-11-09" already exists
Cell In[94], line 1
```

doug-leasure commented on src/covariates/collection/pwtt.py:
```diff
@@ -0,0 +1,106 @@
+country = 'ua'
+
+# data input
+boundaries_oblast_path = repo_dir / 'data' / 'COD-AB' / 'ukr_admbnda_sspe_20230201_SHP'/'ukr_admbnda_adm1_sspe_20230201.shp'
```
Should be lower case cod-ab

edarin replied:
See commit: https://github.com/OxfordDemSci/UkraineNowPop/commit/912846377c4bd0e3c42584c93f923ea85a699ff7

doug-leasure reviewed (commented):
This is coming together really nicely! Very well organised. 

It would be good to have a look through old covariates code to harmonise path names. See [airwars.py](https://github.com/OxfordDemSci/UkraineNowPop/blob/covariates/src/covariates/collection/airwars.py#L8) for an example of a path that needs to be updated.

I will have some edits for deepstate.py and still need to see if I can replicate pwtt, sirens, and warfires. Gotta go for the evening, so just leaving a few comments here already.

edarin committed:
import net_friction

edarin committed:
std path name

doug-leasure committed:
rearrange deepstate script

doug-leasure commented:
I worked on the deepstate.py script, but I could not successfully create the plot at the end of the original script or the revised script. So, I kept both scripts and named the new one deepstate_alt.py in my last commit. @edarin , let's revisit next week to see if we can figure out the problem. I documented a few errors in my review comments that may or may not be relevant.

doug-leasure committed:
pip install net_friction

doug-leasure committed:
cleanup acled script

doug-leasure commented on src/covariates/collection/acled.py:
```diff
@@ -1,58 +1,162 @@
+    for buffer in [1000, 5000, 10000]:
+        vary_buffer_in_subset(
+            buffer_distance_in_meters=buffer,
+            edges=out_dir / "covariates" / "network_edges_oblast.gpkg",
```
This file `.../covariates/network_edges_oblast.gpkg` is not created by this script, so this part of the code fails. Can we add the creation of this file into the script?

doug-leasure replied:
I see now that `./processing/20_create_network.py` must be run before this chunk of code will work. I would suggest removing this bit from `./collection/acled.py` and moving it over to the processing folder. Maybe it is actually redundant with code already in `./processing/21_main.py`?

edarin replied:
It is not redundant. It is a preprocessing step to filter out ACLED data in buffer around corridors before counting the number of acled points. But I will remove it as we are not working with describing corridor yet. and it is currently a clunky integration of David's code.

doug-leasure committed:
update pwtt.py to run on other computers

doug-leasure committed:
install dependencies of sirens.R

doug-leasure reviewed (changes requested):
I have now run most of the covariates pipeline locally and committed changes along the way that were required to make it work on my two machines (Windows and Linux). I left a few comments on specific pieces of code that need your attention. I marked comments as resolved when I was able to fix them myself. A few general comments below...

doug-leasure committed:
lingering unsaved commit

edarin committed:
std name in pwtt as we can deploy it in other contexts

doug-leasure commented:
I created a deepstate_alt.py that I think we should use to replace the current script (which I also modified slightly to make it work). Please check deepstate_alt.py to make sure it works as intended and then rename it to replace the other script if you like it. I couldn't replicate the plot at the end with either script on either of my machines.

doug-leasure commented:
I think we need to be more consistent about defining NAs in the interim data so that we don't impute data in later steps that should actually be zeros:
- deepstate occupied area: NAs should be changed to zeros because it is a national dataset. Anywhere not mapped as occupied is de facto mapped as unuoccupied.
- pwtt building damage: For dates with imagery, NAs should be zeros because the measure is based on images with national coverage. For dates with no imagery, the values should be NA.
- sirens: dates/locations with no data should be included as zeros because it is a national dataset. Currently dates/locations with no sirens are omitted from the interim data completely (i.e. not even included as NA).

doug-leasure commented:
I am supportive of the wide format that you've chosen for the final covariates, but I am a little worried that this will become unwieldy later as we add more covariates. We may want to consider revising the column naming in all interim covariate tables to include a "x_value" column that contains the covariate value, and an "x_name" column that includes the covariate name repeated for every row. With this change, we would have the option to rbind the covariate tables into one long-format dataset, if needed, while preserving the option to continue using wide-format for now (although I know that this would require some changes to 10_combine_sockVariables.R). What do you think?

doug-leasure commented:
PCODES and NAMES should probably be added to the master index (if they aren't already) like you've included them in the covariate tables.

edarin committed:
change deepstate aggregation method from sum to mean ie the mean occupied area during the week

edarin committed:
take _alt as the correct script for deepstate

edarin committed:
replace na by 0 in sirens

edarin committed:
in pwtt replace nan by 0 when imaegry date exists

edarin commented:
> I think we need to be more consistent about defining NAs in the interim data so that we don't impute data that should actually be zeros in later steps:
> 
>     * deepstate occupied area: NAs should be changed to zeros because it is a national dataset. Anywhere not mapped as occupied is de facto mapped as unuoccupied.
> 
>     * pwtt building damage: For dates with imagery, NAs should be zeros because the measure is based on images with national coverage. For dates with no imagery, the values should be NA.
> 
>     * sirens: dates/locations with no data should be included as zeros because it is a national dataset. Currently dates/locations with no sirens are omitted from the interim data completely (i.e. not even included as NA).

Agreed. Implemented. 
I am surprised for sirens as on my side (dates,locations) with no sirens data were already included.

edarin commented:
> PCODES and NAMES should probably be added to the master index (if they aren't already) like you've included them in the covariate tables.

PCODE already included. NAMES not as laready using the geo_name from fb in i_name.

edarin commented:
> I created a deepstate_alt.py that I think we should use to replace the current script (which I also modified slightly to make it work). Please check deepstate_alt.py to make sure it works as intended and then rename it to replace the other script if you like it. I couldn't replicate the plot at the end with either script on either of my machines.

Fixed. I also changed the aggregate of occupied areas through the week as `mean` rather than `sum`. Correct?

edarin committed:
remove corridor description from acled

edarin committed:
Merge branch 'dev' into covariates

doug-leasure commented:
> > PCODES and NAMES should probably be added to the master index (if they aren't already) like you've included them in the covariate tables.
> 
> PCODE already included. NAMES not as laready using the geo_name from fb in i_name. 

I think we should switch to the pcode names for i_name as they take political sensitivities into account. For example, Meta uses the Russian spelling of Kiev whereas UN uses the Ukrainian Kyiv. I think I changed this obvious one manually, but not sure about others so best to follow UN rather than Meta.

doug-leasure commented:
> > I created a deepstate_alt.py that I think we should use to replace the current script (which I also modified slightly to make it work). Please check deepstate_alt.py to make sure it works as intended and then rename it to replace the other script if you like it. I couldn't replicate the plot at the end with either script on either of my machines.
> 
> Fixed. I also changed the aggregate of occupied areas through the week as `mean` rather than `sum`. Correct?

I trust your judgement. Mean makes sense to me.

doug-leasure reviewed (approved):
Thanks very much. I responded to your individual comments. Go ahead and merge when ready.

edarin committed:
swap meta name for cod-ab name in i_name

edarin commented:
> > > PCODES and NAMES should probably be added to the master index (if they aren't already) like you've included them in the covariate tables.
> > 
> > 
> > PCODE already included. NAMES not as laready using the geo_name from fb in i_name.
> 
> I think we should switch to the pcode names for i_name as they take political sensitivities into account. For example, Meta uses the Russian spelling of Kiev whereas UN uses the Ukrainian Kyiv. I think I changed this obvious one manually, but not sure about others so best to follow UN rather than Meta.

Done! https://github.com/OxfordDemSci/UkraineNowPop/pull/75/commits/1ffc53b76d4de89bf23c3fdb7f557b467dcfeb87
